crates/matrix-sdk/src/send_queue.rs

```diff
@@ -353,30 +366,68 @@ impl RoomSendQueue {
                 }
 
                 Err(err) => {
-                    warn!(txn_id = %queued_event.transaction_id, "error when sending event: {err}");
-
-                    // In this case, we intentionally keep the event in the queue, but mark it as
-                    // not being sent anymore.
-                    queue.mark_as_not_being_sent(&queued_event.transaction_id).await;
-
-                    // Let observers know about a failure *after* we've marked the item as not
-                    // being sent anymore. Otherwise, there's a possible race where a caller might
-                    // try to remove an item, while it's still marked as being sent, resulting in a
-                    // cancellation failure.
-
-                    // Disable the queue for this room after an error.
-                    locally_enabled.store(false, Ordering::SeqCst);
+                    let (err, is_recoverable) = if let crate::Error::Http(http_err) = err {
+                        // A small trick to confuse borrowck (and humans too).
+                        //
+                        // `is_transient_error` really wants ownership of the error, so we can't
+                        // match above on `&err`, and we have to pass the error around by ownership
+                        // all over the place. We give ownership in `is_transient_error`, and then
+                        // get it back with the below call to `.error()`. Oh well.
+                        let retry_kind = characterize_retry_kind(http_err);
+
+                        let mut is_recoverable = matches!(retry_kind, RetryKind::Transient { .. });
+
+                        let http_err = retry_kind.error();
+
+                        if !is_recoverable {
+                            // Also flag all HTTP errors as recoverable: they might indicate that
+                            // the server was unreachable, or that the device is entirely
+                            // disconnected.
+                            is_recoverable =
+                                matches!(http_err, crate::error::HttpError::Reqwest(..));
+                        }
+
+                        (crate::Error::Http(http_err), is_recoverable)
+                    } else {
+                        (err, false)
+                    };
+
+                    if is_recoverable {
```

Comment:
Can't we move the whole `is_recoverable` logic under the `HttpError`? This way this becomes:

```rust
if http_err.is_recoverable() {
```

Which probably makes it easier to define what a recoverable error is and test as well.

Reply:
I didn't introduce a new method as now checking for a recoverable error really is checking that the RetryKind is Transient :smiling_face_with_three_hearts: